lib/srv/desktop/discovery.go

```diff
@@ -66,15 +66,24 @@ func (s *WindowsService) startDesktopDiscovery(ctx context.Context) error {
 
 		GetCurrentResources: func() types.ResourcesWithLabels { return s.lastDiscoveryResults },
 		GetNewResources:     s.getDesktopsFromLDAP,
-		OnCreate:            s.createDesktop,
-		OnUpdate:            s.updateDesktop,
-		OnDelete:            s.deleteDesktop,
-		Log:                 s.cfg.Log,
+		// On the first run of the reconcile loop, a desktop may exist in the database
+		OnCreate: s.createDesktop,
+		OnUpdate: s.updateDesktop,
+		OnDelete: s.deleteDesktop,
+		Log:      s.cfg.Log,
 	})
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
+	initialDesktops, err := s.getDesktopsFromAuthServer(ctx)
```

Comment:
Did you consider/try using upsert instead of create/update? It may avoid the need for this call.

Reply:
Yeah. In my opinion this implementation is better because it ensures we never do anything weird on the first run that breaks the semantics of the `Reconciler`.

 The example I thought of is if we have a desktop in the database but are starting with an empty `s.lastDiscoveryResults`, then "discover" (via LDAP) the desktop we have in the database on the first loop, but this time it has different labels. In that case with the upsert design, the `Reconciler` would call it's `OnCreate` (upsert), but really it would be doing an update and overwriting an existing resource. On the other hand, with this design it would accurately call `OnUpdate`.

I'm pretty sure this is immaterial at this point, but am arguing that all else equal it's preferable to stay disciplined with our object's semantics.